File: app/src/main/java/be/digitalia/fosdem/api/LiveRoomStatusesLiveData.java
package be.digitalia.fosdem.api;

import android.annotation.SuppressLint;
import android.os.*;
import android.text.format.DateUtils;
import androidx.lifecycle.LiveData;
import be.digitalia.fosdem.model.RoomStatus;
import be.digitalia.fosdem.parsers.RoomStatusesParser;
import be.digitalia.fosdem.utils.HttpUtils;

import java.io.InputStream;
import java.util.Collections;
import java.util.Map;

/**
 * Loads and maintain the Room statuses live during the event.
 */
class LiveRoomStatusesLiveData extends LiveData<Map<String, RoomStatus>> {

	private static final long REFRESH_DELAY = 90L * DateUtils.SECOND_IN_MILLIS;
	private static final long FIRST_ERROR_REFRESH_DELAY = 30L * DateUtils.SECOND_IN_MILLIS;
	private static final long EXPIRATION_DELAY = 6L * DateUtils.MINUTE_IN_MILLIS;

	private static final int EXPIRE_WHAT = 0;
	private static final int REFRESH_WHAT = 1;

	private final Handler handler = new Handler(Looper.getMainLooper(), new Handler.Callback() {
		@Override
		public boolean handleMessage(Message msg) {
			switch (msg.what) {
				case EXPIRE_WHAT:
					expire();
					return true;
				case REFRESH_WHAT:
					refresh();
					return true;
			}
			return false;
		}
	});

	private long expirationTime = Long.MAX_VALUE;
	private long nextRefreshTime = 0L;
	private int retryAttempt = 0;
	private AsyncTask<Void, Void, Map<String, RoomStatus>> currentTask = null;

	@Override
	protected void onActive() {
		long now = SystemClock.elapsedRealtime();
		if (expirationTime != Long.MAX_VALUE) {
			if (now < expirationTime) {
				handler.sendEmptyMessageDelayed(EXPIRE_WHAT, expirationTime - now);
			} else {
				expire();
			}
		}
		if (now < nextRefreshTime) {
			handler.sendEmptyMessageDelayed(REFRESH_WHAT, nextRefreshTime - now);
		} else {
			refresh();
		}
	}

	@Override
	protected void onInactive() {
		handler.removeMessages(EXPIRE_WHAT);
		handler.removeMessages(REFRESH_WHAT);
	}

	@SuppressLint("StaticFieldLeak")
	void refresh() {
		if (currentTask != null) {
			// Let the ongoing task complete with success or error
			return;
		}
		currentTask = new AsyncTask<Void, Void, Map<String, RoomStatus>>() {

			@Override
			protected Map<String, RoomStatus> doInBackground(Void... voids) {
				try {
					InputStream is = HttpUtils.get(FosdemUrls.getRooms());
					try {
						return new RoomStatusesParser().parse(is);
					} finally {
						is.close();
					}
				} catch (Throwable e) {
					return null;
				}
			}

			@Override
			protected void onPostExecute(Map<String, RoomStatus> result) {
				currentTask = null;
				if (result != null) {
					onSuccess(result);
				} else {
					onError();
				}
			}
		};
		currentTask.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
	}

	void onSuccess(Map<String, RoomStatus> result) {
		setValue(result);
		retryAttempt = 0;
		long now = SystemClock.elapsedRealtime();
		expirationTime = now + EXPIRATION_DELAY;
		if (hasActiveObservers()) {
			handler.sendEmptyMessageDelayed(EXPIRE_WHAT, EXPIRATION_DELAY);
		}
		scheduleNextRefresh(now, REFRESH_DELAY);
	}

	void onError() {
		// Use exponential backoff for retries
		long multiplier = (long) Math.pow(2, retryAttempt);
		retryAttempt++;
		scheduleNextRefresh(SystemClock.elapsedRealtime(),
				Math.min(FIRST_ERROR_REFRESH_DELAY * multiplier, REFRESH_DELAY));
	}

	private void scheduleNextRefresh(long now, long delay) {
		nextRefreshTime = now + delay;
		if (hasActiveObservers()) {
			handler.sendEmptyMessageDelayed(REFRESH_WHAT, delay);
		}
	}

	void expire() {
		// When the data expires, replace it with an empty value
		setValue(Collections.<String, RoomStatus>emptyMap());
		expirationTime = Long.MAX_VALUE;
	}
}

File: app/src/main/java/be/digitalia/fosdem/api/LiveRoomStatusesLiveData.kt
package be.digitalia.fosdem.api

import android.annotation.SuppressLint
import android.os.*
import android.text.format.DateUtils
import androidx.lifecycle.LiveData
import be.digitalia.fosdem.api.FosdemUrls.rooms
import be.digitalia.fosdem.model.RoomStatus
import be.digitalia.fosdem.parsers.RoomStatusesParser
import be.digitalia.fosdem.utils.network.HttpUtils.get
import kotlin.math.pow

/**
 * Loads and maintain the Room statuses live during the event.
 */
internal class LiveRoomStatusesLiveData : LiveData<Map<String, RoomStatus>>() {

    private val handler = Handler(Looper.getMainLooper(), Handler.Callback { msg: Message ->
        when (msg.what) {
            EXPIRE_WHAT -> {
                expire()
                true
            }
            REFRESH_WHAT -> {
                refresh()
                true
            }
            else -> false
        }
    })
    private var expirationTime = Long.MAX_VALUE
    private var nextRefreshTime = 0L
    private var retryAttempt = 0
    // TODO replace with LiveData coroutines builder
    private var currentTask: AsyncTask<Unit, Nothing, Map<String, RoomStatus>?>? = null

    override fun onActive() {
        val now = SystemClock.elapsedRealtime()
        if (expirationTime != Long.MAX_VALUE) {
            if (now < expirationTime) {
                handler.sendEmptyMessageDelayed(EXPIRE_WHAT, expirationTime - now)
            } else {
                expire()
            }
        }
        if (now < nextRefreshTime) {
            handler.sendEmptyMessageDelayed(REFRESH_WHAT, nextRefreshTime - now)
        } else {
            refresh()
        }
    }

    override fun onInactive() {
        handler.removeMessages(EXPIRE_WHAT)
        handler.removeMessages(REFRESH_WHAT)
    }

    @SuppressLint("StaticFieldLeak")
    fun refresh() {
        if (currentTask != null) {
            // Let the ongoing task complete with success or error
            return
        }
        val task = object : AsyncTask<Unit, Nothing, Map<String, RoomStatus>?>() {

            override fun doInBackground(vararg params: Unit): Map<String, RoomStatus>? {
                return try {
                    get(rooms).use { source -> RoomStatusesParser().parse(source) }
                } catch (e: Throwable) {
                    null
                }
            }

            override fun onPostExecute(result: Map<String, RoomStatus>?) {
                currentTask = null
                if (result != null) {
                    onSuccess(result)
                } else {
                    onError()
                }
            }
        }
        currentTask = task
        task.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR)
    }

    private fun onSuccess(result: Map<String, RoomStatus>) {
        value = result
        retryAttempt = 0
        val now = SystemClock.elapsedRealtime()
        expirationTime = now + EXPIRATION_DELAY
        if (hasActiveObservers()) {
            handler.sendEmptyMessageDelayed(EXPIRE_WHAT, EXPIRATION_DELAY)
        }
        scheduleNextRefresh(now, REFRESH_DELAY)
    }

    private fun onError() {
        // Use exponential backoff for retries
        val multiplier = 2.0.pow(retryAttempt).toLong()
        retryAttempt++
        scheduleNextRefresh(SystemClock.elapsedRealtime(),
                (FIRST_ERROR_REFRESH_DELAY * multiplier).coerceAtMost(REFRESH_DELAY))
    }

    private fun scheduleNextRefresh(now: Long, delay: Long) {
        nextRefreshTime = now + delay
        if (hasActiveObservers()) {
            handler.sendEmptyMessageDelayed(REFRESH_WHAT, delay)
        }
    }

    private fun expire() {
        // When the data expires, replace it with an empty value
        value = emptyMap()
        expirationTime = Long.MAX_VALUE
    }

    companion object {
        private const val REFRESH_DELAY = 90L * DateUtils.SECOND_IN_MILLIS
        private const val FIRST_ERROR_REFRESH_DELAY = 30L * DateUtils.SECOND_IN_MILLIS
        private const val EXPIRATION_DELAY = 6L * DateUtils.MINUTE_IN_MILLIS
        private const val EXPIRE_WHAT = 0
        private const val REFRESH_WHAT = 1
    }
}